### ISSUE TITLE
df and f method of loss function class?

### ISSUE THREAD

**Issue by kminsu1204** [tiny-dnn/tiny-dnn#1000]
For example, I can understand the mean squared error equation from f method, but cannot understand logic of df method. Can anyone explain it to me?

`class mse {
 public:
  static float_t f(const vec_t &y, const vec_t &t) {
    assert(y.size() == t.size());
    float_t d{0.0};

    for (size_t i = 0; i < y.size(); ++i) d += (y[i] - t[i]) * (y[i] - t[i]);

    return d / static_cast<float_t>(y.size());
  }

  static vec_t df(const vec_t &y, const vec_t &t) {
    assert(y.size() == t.size());
    vec_t d(t.size());
    float_t factor = float_t(2) / static_cast<float_t>(t.size());

    for (size_t i = 0; i < y.size(); ++i) d[i] = factor * (y[i] - t[i]);

    return d;
  }
};`

**Comment by kminsu1204** [comment]
I figured out by myself that f method is used to get loss using mse equation itself, and df is used for gradient descent, so it's a derivative of above one.
